File: plugins/bigtiny/bigtiny/server/routes/providers.py
from __future__ import annotations

import json
import logging
from uuid import uuid4

import keyring
from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/providers")


class CreateProviderRequest(BaseModel):
    name: str
    provider_type: str  # "openai_compat" | "anthropic"
    base_url: str
    api_key: str | None = None
    fallback_priority: int = 1
    config: dict | None = None


class UpdateProviderRequest(BaseModel):
    name: str | None = None
    base_url: str | None = None
    api_key: str | None = None
    fallback_priority: int | None = None
    config: dict | None = None


@router.get("")
async def list_providers(request: Request):
    db = request.app.state.db
    rows = await db.fetch_all("SELECT * FROM providers ORDER BY fallback_priority")
    return {"providers": [dict(r) for r in rows]}


@router.post("")
async def add_provider(body: CreateProviderRequest, request: Request):
    db = request.app.state.db
    provider_id = uuid4().hex[:8]
    await db.execute(
        "INSERT INTO providers (id, name, provider_type, base_url, fallback_priority, config) "
        "VALUES (:id, :name, :type, :url, :priority, :config)",
        {
            "id": provider_id,
            "name": body.name,
            "type": body.provider_type,
            "url": body.base_url,
            "priority": body.fallback_priority,
            "config": json.dumps(body.config or {}),
        },
    )
    if body.api_key:
        keyring.set_password("bigtiny", f"{provider_id}_api_key", body.api_key)

    router = request.app.state.router
    row = {
        "id": provider_id,
        "name": body.name,
        "provider_type": body.provider_type,
        "base_url": body.base_url,
        "fallback_priority": body.fallback_priority,
        "config": body.config or {},
        "status": "disconnected",
    }
    await router.register_provider(row, body.api_key)

    return {"id": provider_id, "status": "created"}


@router.patch("/{provider_id}")
async def update_provider(provider_id: str, body: UpdateProviderRequest, request: Request):
    db = request.app.state.db
    existing = await db.fetch_one(
        "SELECT * FROM providers WHERE id = :id", {"id": provider_id}
    )
    if not existing:
        raise HTTPException(404, "Provider not found")

    updates: dict[str, object] = {}
    if body.name is not None:
        updates["name"] = body.name
    if body.base_url is not None:
        updates["base_url"] = body.base_url
    if body.fallback_priority is not None:
        updates["fallback_priority"] = body.fallback_priority
    if body.config is not None:
        updates["config"] = json.dumps(body.config)

    if updates:
        set_clause = ", ".join(f"{k} = :{k}" for k in updates)
        updates["id"] = provider_id
        await db.execute(
            f"UPDATE providers SET {set_clause}, updated_at = CURRENT_TIMESTAMP WHERE id = :id",
            updates,
        )

    if body.api_key:
        keyring.set_password("bigtiny", f"{provider_id}_api_key", body.api_key)

    router = request.app.state.router
    row = await db.fetch_one(
        "SELECT * FROM providers WHERE id = :id", {"id": provider_id}
    )
    if row:
        await router.register_provider(dict(row), body.api_key)

    return {"status": "updated"}


@router.delete("/{provider_id}")
async def delete_provider(provider_id: str, request: Request):
    db = request.app.state.db
    await db.execute("DELETE FROM providers WHERE id = :id", {"id": provider_id})
    try:
        keyring.delete_password("bigtiny", f"{provider_id}_api_key")
    except Exception:
        pass

    router = request.app.state.router
    router.unregister_provider(provider_id)

    return {"status": "deleted"}


@router.post("/{provider_id}/test")
async def test_provider(provider_id: str, request: Request):
    router = request.app.state.router
    provider = router._providers.get(provider_id)
    if not provider:
        raise HTTPException(404, "Provider not found")
    try:
        models = await provider.discover_models()
        return {
            "status": "connected",
            "models": [m.model_dump() for m in models],
        }
    except Exception as e:
        return {"status": "error", "error": str(e)}


@router.get("/{provider_id}/models")
async def list_models(provider_id: str, request: Request):
    router = request.app.state.router
    provider = router._providers.get(provider_id)
    if not provider:
        raise HTTPException(404, "Provider not found")
    models = await provider.discover_models()
    return {"models": [m.model_dump() for m in models]}

File: plugins/bigtiny_rust/src/storage/providers.rs
use chrono::DateTime;
use chrono::Utc;
use serde::{Deserialize, Serialize};
use sqlx::{FromRow, SqlitePool};

use crate::error::StorageError;

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
pub struct ProviderRow {
    pub id: String,
    pub name: String,
    pub provider_type: String,
    pub base_url: String,
    pub fallback_priority: i32,
    pub config: Option<String>,
    pub status: String,
    pub error_message: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

pub async fn list_providers(pool: &SqlitePool) -> Result<Vec<ProviderRow>, StorageError> {
    let rows = sqlx::query_as::<_, ProviderRow>(
        r#"SELECT id, name, provider_type, base_url, fallback_priority, config, status, error_message, created_at, updated_at
           FROM providers ORDER BY fallback_priority ASC"#
    )
    .fetch_all(pool)
    .await?;
    Ok(rows)
}

pub async fn get_provider(
    pool: &SqlitePool,
    provider_id: &str,
) -> Result<Option<ProviderRow>, StorageError> {
    let row = sqlx::query_as::<_, ProviderRow>(
        r#"SELECT id, name, provider_type, base_url, fallback_priority, config, status, error_message, created_at, updated_at
           FROM providers WHERE id = ?"#
    )
    .bind(provider_id)
    .fetch_optional(pool)
    .await?;
    Ok(row)
}

pub async fn create_provider(
    pool: &SqlitePool,
    id: &str,
    name: &str,
    provider_type: &str,
    base_url: &str,
) -> Result<ProviderRow, StorageError> {
    sqlx::query(r#"INSERT INTO providers (id, name, provider_type, base_url) VALUES (?, ?, ?, ?)"#)
        .bind(id)
        .bind(name)
        .bind(provider_type)
        .bind(base_url)
        .execute(pool)
        .await?;
    get_provider(pool, id)
        .await?
        .ok_or_else(|| StorageError::Generic(format!("Provider {} not found after creation", id)))
}

pub async fn update_provider(
    pool: &SqlitePool,
    provider_id: &str,
    name: Option<&str>,
    base_url: Option<&str>,
    config: Option<&str>,
) -> Result<(), StorageError> {
    sqlx::query(
        r#"UPDATE providers SET name = COALESCE(?1, name), base_url = COALESCE(?2, base_url),
           config = COALESCE(?3, config), updated_at = CURRENT_TIMESTAMP WHERE id = ?4"#,
    )
    .bind(name)
    .bind(base_url)
    .bind(config)
    .bind(provider_id)
    .execute(pool)
    .await?;
    Ok(())
}

pub async fn delete_provider(pool: &SqlitePool, provider_id: &str) -> Result<u64, StorageError> {
    let result = sqlx::query(r#"DELETE FROM providers WHERE id = ?"#)
        .bind(provider_id)
        .execute(pool)
        .await?;
    Ok(result.rows_affected())
}

pub async fn update_provider_status(
    pool: &SqlitePool,
    provider_id: &str,
    status: &str,
    error_message: Option<&str>,
) -> Result<(), StorageError> {
    sqlx::query(
        r#"UPDATE providers SET status = ?, error_message = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?"#
    )
    .bind(status)
    .bind(error_message)
    .bind(provider_id)
    .execute(pool)
    .await?;
    Ok(())
}
